Improve Prefab Overrides Docs (#2506)
o3de/o3de.org: content/docs/learning-guide/tutorials/entities-and-prefabs/override-a-prefab.md

| O3DE Experience | Time to Complete | Feature Focus | Last Updated |
| --- | --- | --- | --- |
| Beginner | 15 Minutes | Using prefab overrides to make prefab instances unique. | November 20, 2023 |

## Prerequisites

* Basic knowledge of working with [O3DE Editor](/docs/user-guide/editor).
* A project built from the standard project template or one that contains the Gems in the standard template.

## What are prefab overrides?

> Prefab overrides are one of the major components of the prefab system. Before overrides, the only way to modify prefabs is by entering Prefab Edit Mode but changes would be applied to all instances of that prefab. With overrides, we can make prefabs stand out from the rest of their siblings, which gives content creators much more control and flexibility in building large and complex scenes. It also opens doors for easily overriding default procedural prefabs that are imported from `.fbx` files.

Author: ```suggestion
Prefab overrides are one of the major components of the prefab system. Before overrides, changing a single prefab instance would affect all instances. With overrides, we can make prefabs stand out from the rest of their siblings. Overrides give content creators more flexibility to build large and complex scenes. It also allows for overriding default procedural prefabs imported via `.fbx` files.
```

The following image shows two instances of Car prefab in a level. With overrides enabled, the Car instances are expandable and can be opened for editing:

{{< image-width src="/images/learning-guide/tutorials/entities-and-prefabs/level-prefab-edit.png" width="300" alt="Level in Prefab Edit Mode in Entity Outliner." >}}

{{< note >}}
If prefab overrides are disabled, prefab instances in the level cannot be expanded for editing. In other words, you will not see the arrow icon to the left of an instance name.
{{< /note >}}

Overrides applied to prefab instances are registered for that individual instance and are stored in the prefab being edited that contains the instances. In the above example, the Car instances are in a level (which is a prefab), so the prefab overrides applied to the Car instances are stored in the level when you save it.

{{< image-width src="/images/learning-guide/tutorials/entities-and-prefabs/level-prefab-edit.png" width="300" alt="Level in Prefab Edit Mode with Override Edit in Entity Outliner." >}}

> Suppose that you modify the Body entity's location as an override from the level prefab, you notice that a blue circle will appear on the entity's icon in Entity Outliner. This indicate that an override has been applied to the entity. Now, if you save the level prefab, the override will be stored.

Author: ```suggestion
Modifying the Body entity's location causes a blue circle to appear in the Entity Outliner. This indicates an entity override. Now, if you save the level prefab, the override is also saved.
```

{{< note >}}
A level is a prefab, and automatically enters Prefab Edit Mode when it is opened. This is indicated by the blue capsule around the level in **Entity Outliner**.

Overrides are not limited to the level. In fact, any prefab that is open for editing in Prefab Edit Mode is responsible for storing the overrides made to its nested prefabs.
{{< /note >}}

> However, if you do not want to make override changes, you should follow the old Prefab Edit Mode workflow instead. In the example, you can enter the mode on the Car instance before editing the Body entity, in which any changes will be applied to all Car instances as shown in the below image.

Author: ```suggestion
If you do not want to make override changes, you should follow the old Prefab Edit Mode workflow instead. In this example, you can enter Prefab Edit Mode via the Car instance in the Entity Outliner. Any changes to Car will now affect all Car instances.
```

Reviewer: Love the paraphrasing!

1. Notice that a blue circle appears on the entity's icon in Entity Outliner. This indicates that an override is applied to the entity.

In the below image, the Body entity in the first Car instance has an override to change the default color of the car from red to blue:

{{< image-width src="/images/learning-guide/tutorials/entities-and-prefabs/prefab-override-property.png" width="750" alt="Overriding a component property." >}}

When you make an override edit in the Inspector, you will notice that a blue circle appears before the property label and the label text is bolded. In the component header, a blue circle is also shown near the text, which indicates that an owning property value has been overridden.

{{< image-width src="/images/learning-guide/tutorials/entities-and-prefabs/prefab-override-component-property.png" width="450" alt="Overriding a component property." >}}

### Add or remove components

To add a component of an entity under a prefab instance:

1. In Entity Outliner, select an entity under an instance that is not in Prefab Edit Mode.
1. In Inspector, **right-click** on empty area and choose **Hi** from the context menu.

> 1. Notice that a blue plus icon appears on the component's icon to the left or the component name.

Author: ```suggestion
1. Notice that a blue plus icon appears on the component's icon to the left of the component name.
```

{{< image-width src="/images/learning-guide/tutorials/entities-and-prefabs/prefab-override-add.png" width="750" alt="Adding an entity as an override." >}}

To delete an entity or a nested prefab instance as an override:
1. In Entity Outliner, **right-click** on an entity or a nested prefab instance and choose **Delete** from the context menu.
1. Notice that a blue circle appears on the parent entity's icon in Entity Outliner. This indicates that a child has been deleted as an override.

Similarly in Entity Outliner, there is no visual indication of an entity or a prefab instance being deleted as an override at this time. You can follow the steps mentioned above to either remove the override in file or to undo the override in Editor.

{{< note >}}
You cannot revert the deletion by reverting overrides on the parent entity. See the GitHub issue [#13437](https://github.com/o3de/o3de/issues/13437) for more details.
{{< /note >}}

### Remove nested prefab instances

In current workflow, it does not support adding a nested prefab instance as an override. However, you can remove a nested prefab instance as an override by following the steps mentioned above for an entity.

{{< note >}}

> It is important to note that deleting or undoing a prefab instance will not actually delete the actual prefab from the hard drive, which is the same when removing a prefab from the Entity Outliner.

Author: ```suggestion
Deleting or undoing a prefab instance will not delete the actual prefab off the hard drive.
```

### Revert some overrides on the component

In addition, you can revert all property overrides that are owned by a component:
1. In Inspector, **right-click** on the blue circle in the component header and choose **Revert Overrides** from the context menu.
1. Notice that all the properties no longer have any indication of having an override.

{{< image-width src="/images/learning-guide/tutorials/entities-and-prefabs/prefab-override-revert-component-override.png" width="450" alt="Reverting a component added as an override." >}}

For a component added as an override, you can revert the addition in the same way mentioned above. Alternatively, you can delete the component with either of the following methods:

1. In Inspector, **right-click** a component and choose **Delete component** from the context menu.
1. With a component selected, press **DEL**.

For a component deleted as an override, you can revert the deletion by reverting all overrides on the selected entity. However, be aware that this would remove other overrides if they are present.

### Revert all overrides on the entity

> Last but not least, you can revert all overrides that is applied on an entity:

Author: ```suggestion
Last but not least, you can revert all overrides applied to an entity:
```

1. In Entity Outliner, **right-click** on an entity with overrides and choose **Revert Overrides** from the context menu.
1. Notice that the entity no longer has any indication of having overrides.

{{< image-width src="/images/learning-guide/tutorials/entities-and-prefabs/prefab-override-revert.png" width="300" alt="Reverting overrides on an entity." >}}

{{< note >}}

> If you see any unexpected overrides showing up, you can try the above methods to revert them. If it does not work out, you can open the prefab file that owns those overrides (under the `Instances` key) and remove the override patches you want.

Author: pro-tip! <3

{{< /note >}}

## Disable prefab overrides

> Prefab overrides are a new feature. It is possible that you may encounter blocking issues that slow down your workflow. You can follow the recommendations below.

Reviewer: Redundant:

```suggestion
Prefab overrides are a new feature. You may encounter issues that slow down or even block your workflow. You can follow the recommendations below to disable the feature in your project.
```
